kubernetes/updater/pkg/controller/deployment.go

```diff
@@ -0,0 +1,127 @@
+)
+
+// DeploymentVersionUpdater Reconciles a deployment by changing its img
+type DeploymentVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+func (r *DeploymentVersionUpdater) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithValues("namespacedname", req.NamespacedName)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// Get the object we are reconciling
+	var obj appsv1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			// log.Info("not found")
+			return ctrl.Result{}, nil
+		}
+		// log.Error(err, "failed to get resource")
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+
+	// Get the current and past version
+	currentVersion, err := getDeploymentVersion(&obj)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *trace.BadParameterError:
+			log.Info("Teleport container found, but failed to get version from the img tag. Will continue and do a version update.")
+		default:
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	image, err := r.Do(ctx, &obj, currentVersion)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *NoNewVersionError:
+			log.Info("Deployment is already up-to-date, not updating.")
+			return requeueLater, nil
+		case *MaintenanceNotTriggeredError:
+			log.Info("No maintenance triggered, not updating.")
+			return requeueLater, nil
+		case *trace.CompareFailedError:
+			log.Info("Image verification failed, not updating.")
+			return requeueLater, trace.Wrap(err)
+		default:
+			log.Error(err, "Unexpected error, not updating.")
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	err = setContainerImageFromPod(&obj.Spec.Template.Spec, teleportContainerName, image.String())
+	if err != nil {
+		return requeueLater, trace.Wrap(err)
+	}
+
+	err = r.Update(ctx, &obj)
+	if err != nil {
+		return requeueNow, trace.Wrap(err)
+	}
+	return requeueLater, nil
+}
+
+func (r *DeploymentVersionUpdater) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+func getDeploymentVersion(deployment *appsv1.Deployment) (string, error) {
+	var current string
+	image, err := getContainerImageFromPod(deployment.Spec.Template.Spec, teleportContainerName)
+	if err != nil {
+		return current, trace.Wrap(err)
+	}
+
+	// TODO: put this in a function and reuse for statefulset
+	imageRef, err := reference.ParseNamed(image)
+	if err != nil {
+		return current, trace.Wrap(err)
+	}
+	taggedImageRef, ok := imageRef.(reference.Tagged)
+	if !ok {
+		return "", trace.BadParameter("imageRef %s is not tagged", imageRef)
+	}
+	current = "v" + taggedImageRef.Tag()
+	if !semver.IsValid(current) {
+		return "", trace.BadParameter("tag %s is not following semver", current)
+	}
+	for _, suffix := range platformSuffixes {
+		current = strings.TrimSuffix(current, suffix)
```

Comment:
Isn't it better to explode current from the end and exclude the latest `-...`?

Reply:
The platform suffix is optional, and from what I know there are other possible suffixes:
- `-debug` for debug images (see hardened images RFD)
- `-fips` ?
- `-ent` ?
- pre-release suffixes `-alpha.1`

I'm not sure how they are composed and if we always end with the platform. For now it seems to be the case. If this were to change I guess we would have to be a bit smarter and split on `-`, pop the platform element and implode the list to get the final version.

Reply:
Got it

Reply:
I'm unresolving because we'll likely get rid of the platform suffix but there are other possible suffixes. I'll wait for @tcsc 's confirmation before moving on with this.

Reply:
According to trent:
- we don't want to put anything in the tag that is not the version for new distroless images (yay!)

Our existing images can have -fips (only for enterprise) and -platform.

```
public.ecr.aws/gravitational/teleport-ent:12.0.1-fips-amd64
```

I suggest we don't make special cases for the legacy images. I don't think anyone will want to run fips at the in the MVP anyway because cloud is not running FIPS-compliant binaries, and this updater is not available in a FIPS flavour.